### PR TITLE
wip - CometNativeScan

### PR DESCRIPTION
**Pull request by parthchandra**
Replaces hacked `CometScan` with `CometNativeScan` so we can switch between original and experimental without rebuilding.
I tried removing a lot of the functions in CometNativeScan but too many test in CometExecSuite fail and I ended up including almost everything back.
Three tests in CometExecScan are still failing.

### CODE REVIEW

File: common/src/main/scala/org/apache/comet/CometConf.scala [hunk 1]
```diff
@@ -77,6 +77,15 @@ object CometConf extends ShimCometConf {
...
     .createWithDefault(true)
 
+  val COMET_FULL_NATIVE_SCAN_ENABLED: ConfigEntry[Boolean] = conf(
+    "spark.comet.native.scan.enabled")
+    .doc(
+      "Whether to enable the fully native scan. When this is turned on, Spark will use Comet to " +
+        "read supported data sources (currently only Parquet is supported natively)." +
+        " By default, this config is true.")
+    .booleanConf
+    .createWithDefault(true)
```
Comment (viirya) [inline common/src/main/scala/org/apache/comet/CometConf.scala:87]: I added a similar config locally. 🙂

Comment (parthchandra) [reply common/src/main/scala/org/apache/comet/CometConf.scala:87]: If it was was something related, you can merge it here. This is an experimental branch. If it is for something different, then we can change our name.

Comment (viirya) [reply common/src/main/scala/org/apache/comet/CometConf.scala]: No, I meant I tried to do similar native scan change into this branch. I was trying to enable DataFusion native Parquet scan yesterday. I found this branch this morning so tried to apply onto this branch.

I will see if any difference can be merged.

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 2]
```diff
@@ -189,6 +189,22 @@ class CometSparkSessionExtensions
...
+                HadoopFsRelation(_, partitionSchema, _, _, _: ParquetFileFormat, _),
+                _: Seq[_],
+                requiredSchema,
+                _,
+                _,
+                _,
+                _,
+                _,
+                _)
+              if CometNativeScanExec.isSchemaSupported(requiredSchema)
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:202]: we could also fall back for bucketed scans and other unsupported cases (can be a separate PR)

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @parthchandra
